functions.sql

```diff
@@ -0,0 +1,92 @@
+/* Additional database functions for openstreetmap-carto */
+
+/* Access functions below adapted from https://github.com/imagico/osm-carto-alternative-colors/tree/591c861112b4e5d44badd108f4cd1409146bca0b/sql/roads.sql */
+
+
+/* Simplified 'yes', 'restricted', 'no', NULL scale for access restriction 
+  'no' is returned if the rendering for highway category does not support 'destination'.
+  NULL is functionally equivalent to 'yes', but indicates the absence of a restriction 
+  rather than a positive access = yes */
+CREATE OR REPLACE FUNCTION carto_int_access(primary_mode text, accesstag text)
+	RETURNS text
+	LANGUAGE SQL
+	IMMUTABLE PARALLEL SAFE
+AS $$
+SELECT
+	CASE
+	WHEN accesstag IN ('yes', 'designated', 'permissive', 'customers') THEN 'yes'
+	WHEN accesstag IN ('destination',  'delivery', 'permit') THEN
```

Comment:
The distinction that `customers` translates to `yes` while `delivery` and `permit` translate to `restricted` is odd.  Is there anything in the use of these values that warrants that distinction?

As far as i can see the delineation between `permit` and `private` is quite unclear.  We currently use `permit` only on POIs in aggregation with `private`.

My own definition of `restricted` would be _any access condition where anyone can use the road/path ad hoc as long as they do it with a specific purpose/intention under their control_:

* destination: With the purpose to reach a specific destination on the road/path.
* delivery: With the purpose to deliver something to a specific destination on the road/path.
* customers: As a customer/with the purpose to become a customer of a business on the road/path.

Reply:
In terms of `customers`, I agree with your logic that `restricted` corresponds to "with a specific purpose only". I was worried that a lot of service road in car parks would be tag with access restrictions and `this would immediately generate an Issue on the lines of "please get rid of these weird grey circles". But I agree that including `customers` in the "restricted" category would be more logical and would provide better mapper feedback (at the risk that mappers start removing `access=customers` on the basis that they don't like the access marking.

I read `permit` as "authorisation required but generally granted", which would fit with the logic of "with a specific purpose". Personally I feel this ought to be clear enough to distinguish from "you need to be somebody special to use this road", and a `restricted` access marking would reinforce this. I feel lumping with `private` makes sense for a 2 state model (free to all vs. not free to all), but less so for a three state model. But we can leave out `permit` if we decide the delineation is too unclear.

Reply:
Please keep in mind that we do not adjust our rendering to what is written on the wiki, we need to consider how tags are actually used and the wiki is only relevant in how far it documents that.

Reply:
Unfortunately we can't really tell how `access=permit` is being used. I But the argument for "intermediate" in a 3-state rendering is that it gives mappers the option of distinguishing between "you can expect to get a permit" and "you have no chance of getting a permit". The distinction is practically useful. But ultimately I'd go with `no` marking over ignoring permit.

Reply:
I have sampled use of `access=permit` a bit and the majority of uses i have seen are cases of `access=private` where the access is probably practically checked through some formal document, but which cannot be obtained by the general public, like for example in security areas on airports, industrial sites, service roads in schools, hospital campuses etc.

Based on that i would think we should either not interpret `access=permit` due to lack of consistent use with the wiki and de facto use clearly being at conflict.  Or aggregate it with `access=private` - which is the de facto meaning in the majority of cases it seems.

If anyone has had a broader look and has evidence that my sampling is not representative please say so.

Reply:
Happy to aggregate `access=permit` with `access=no` on this basis.

Reply:
Sounds good, `customers` goes to `restricted`, `permit` goes to `no`.